python/port/modturtle_impl.cpp

```diff
@@ -0,0 +1,302 @@
+constexpr int t_x_offset = Ion::Display::Width / 2;
+constexpr int t_y_offset = (Ion::Display::Height-18) / 2;
+constexpr int t_size = 9;
+constexpr int t_icons = 8;
+constexpr KDPoint t_icon_offset(-t_size/2 + 1, -t_size/2 + 1);
+constexpr KDSize t_icon_size(t_size, t_size);
+
+static KDColor t_color;
+static float t_heading;
+static float t_x, t_y;
+static bool t_penup;
+static bool t_hidden;
+static int t_speed;
+static int t_dotsize;
+
+static int t_mileage;
+static bool t_drawn;
+
+static KDColor *t_underneath;
+static KDColor *t_icon;
+static uint8_t *t_dot;
+
+static KDPoint pos_turtle(float x, float y) {
+  return KDPoint(round(x + t_x_offset), round(y + t_y_offset));
+}
+
+void draw_turtle() {
+  MicroPython::ExecutionEnvironment::currentExecutionEnvironment()->displaySandbox();
+
+  int frame = (int)((t_heading / (2*M_PI)) * t_icons + 0.5);
+  if (frame < 0) {
+    frame = t_icons - ((-frame) % t_icons) - 1;
+  }
+  else {
+    frame = frame % t_icons;
+  }
+
+  int offset = frame * (t_size*t_size);
+
+  if (!t_hidden) {
+    KDIonContext::sharedContext()->fillRectWithPixels(KDRect(pos_turtle(t_x, t_y).translatedBy(t_icon_offset), t_icon_size), &t_icon[offset], nullptr, t_drawn ? nullptr : t_underneath);
+    t_drawn = true;
+  }
+
+  if (t_mileage > 1000) {
+    if (t_speed > 0) {
+      Ion::msleep(8 * (8 - t_speed));
+      t_mileage -= 1000;
+    }
+    else {
+      t_mileage = 0;
+    }
+  }
+}
+
+void erase_turtle() {
+  if (t_drawn) {
+    KDIonContext::sharedContext()->fillRectWithPixels(KDRect(pos_turtle(t_x, t_y).translatedBy(t_icon_offset), t_icon_size), t_underneath, nullptr);
+    t_drawn = false;
+  }
+}
+
+void dot_turtle(float x, float y) {
+  MicroPython::ExecutionEnvironment::currentExecutionEnvironment()->displaySandbox();
+
+  if (!t_penup) {
+    KDColor colors[t_dotsize*t_dotsize];
+    KDRect rect(pos_turtle(x, y).translatedBy(KDPoint(-t_dotsize/2, -t_dotsize/2)), KDSize(t_dotsize, t_dotsize));
+    KDIonContext::sharedContext()->blendRectWithMask(rect, t_color, t_dot, colors);
+  }
+
+  if (t_speed > 0) {
+    t_mileage += sqrt((x - t_x) * (x - t_x) + (y - t_y) * (y - t_y)) * 1000;
+  }
+
+  micropython_port_should_interrupt();
+
+  t_x = x;
+  t_y = y;
+}
+
+mp_obj_t turtle_forward(mp_obj_t px) {
+  float x = t_x + mp_obj_get_float(px)*sin(t_heading);
+  float y = t_y + mp_obj_get_float(px)*cos(t_heading);
+  turtle_moveto(x, y);
+  return mp_const_none;
+}
+
+mp_obj_t turtle_backward(mp_obj_t px) {
+  return turtle_forward(mp_obj_new_float(-mp_obj_get_float(px)));
+}
+
+mp_obj_t turtle_right(mp_obj_t angle) {
+  return turtle_left(mp_obj_new_float(-mp_obj_get_float(angle)));
+}
+
+mp_obj_t turtle_left(mp_obj_t angle) {
+  float new_angle = ((t_heading - t_heading_offset) + (mp_obj_get_float(angle) * t_heading_scale)) / t_heading_scale;
+  return turtle_setheading(mp_obj_new_float(new_angle));
+}
+
+mp_obj_t turtle_goto(size_t n_args, const mp_obj_t *args) {
+  float newx, newy;
```

Comment:
Use mp_float_t which is a dynamic type. MicroPythons floats can either be single or double precision floats, and in our case they are actually doubles.
```suggestion
  mp_float_t newx, newy;
```

Reply:
Also, we might want to remove single-precision floats altogether at some point, so unless performance is critical I suggest using doubles when possible.